pkg/plugins/mellanox/mellanox_plugin.go

```diff
@@ -71,42 +73,72 @@ func (p *MellanoxPlugin) OnNodeStateChange(old, new *sriovnetworkv1.SriovNetwork
 	needReboot = false
 	err = nil
 	attributesToChange = map[string]mlnxNic{}
+	mellanoxNicsStatus = map[string]map[string]*sriovnetworkv1.InterfaceExt{}
+	mellanoxNicsSpec = map[string]*sriovnetworkv1.Interface{}
+	processedNics := map[string]bool{}
 
+	for _, iface := range new.Status.Interfaces {
+		if iface.Vendor != MellanoxVendorId {
+			continue
+		}
+
+		pciPrefix := getPciAddressPrefix(iface.PciAddress)
+		if ifaces, ok := mellanoxNicsStatus[pciPrefix]; ok {
+			ifaces[iface.PciAddress] = &iface
+		} else {
+			mellanoxNicsStatus[pciPrefix] = map[string]*sriovnetworkv1.InterfaceExt{iface.PciAddress: &iface}
+		}
+	}
+
+	// Add only mellanox cards that required changes in the map, to help track dual port NICs
 	for _, iface := range new.Spec.Interfaces {
-		if !isMlnxNicAndInNode(iface.PciAddress, new) {
+		pciPrefix := getPciAddressPrefix(iface.PciAddress)
+		if _, ok := mellanoxNicsStatus[pciPrefix]; !ok {
+			continue
+		}
+
+		mellanoxNicsSpec[iface.PciAddress] = &iface
+	}
+
+	for _, ifaceSpec := range mellanoxNicsSpec {
+		pciPrefix := getPciAddressPrefix(ifaceSpec.PciAddress)
+		// skip processed nics, help not running the same logic 2 times for dual port NICs
+		if _, ok := processedNics[pciPrefix]; ok {
 			continue
 		}
-		fwCurrent, fwNext, err := getMlnxNicFwData(iface.PciAddress)
+		processedNics[pciPrefix] = true
+
+		fwCurrent, fwNext, err := getMlnxNicFwData(ifaceSpec.PciAddress)
 		if err != nil {
 			return false, false, err
 		}
 		if fwCurrent.enableSriov != fwNext.enableSriov || fwCurrent.totalVfs != fwNext.totalVfs ||
-			fwCurrent.linkType != fwNext.linkType {
+			fwCurrent.linkTypeP1 != fwNext.linkTypeP1 || fwCurrent.linkTypeP2 != fwNext.linkTypeP2 {
 			needReboot = true
-			break
+			continue
 		}
-		attrs := &mlnxNic{totalVfs: -1, firstPort: fwCurrent.firstPort}
-		requireChange := false
 
-		if fwCurrent.totalVfs != iface.NumVfs {
-			attrs.totalVfs = iface.NumVfs
-			requireChange = true
-		}
-		if iface.NumVfs == 0 && fwCurrent.enableSriov {
-			attrs.enableSriov = false
-			requireChange = true
-		} else if iface.NumVfs > 0 && !fwCurrent.enableSriov {
-			attrs.enableSriov = true
-			requireChange = true
-		}
-		if fwCurrent.linkType != EthLinkType && fwCurrent.linkType != PreconfiguredLinkType && fwCurrent.linkType != UknownLinkType {
-			attrs.linkType = EthLinkType
-			requireChange = true
+		isDualPort := isDualPort(ifaceSpec.PciAddress)
+		otherIfaceSpec := getOtherPortSpec(ifaceSpec.PciAddress)
+		otherIfaceStatus := getOtherPortStatus(ifaceSpec.PciAddress)
+		// Initialing the attributes to change
+		attrs := &mlnxNic{enableSriov: false, totalVfs: -1, linkTypeP1: fwCurrent.linkTypeP1,
+			linkTypeP2: fwCurrent.linkTypeP2}
+
+		var totalVfs int
+		totalVfs, needReboot = handleTotalVfs(fwCurrent, attrs, ifaceSpec, otherIfaceSpec, otherIfaceStatus)
```

Comment:
Mellanox plugin changes the total VFs (which needs reboots) to be the max number of request VFs, for single port it must do reboot on change of vfs, for dual port it will trigger reboot if the max is changed